src/coupled_driver.cpp

```diff
@@ -267,216 +283,277 @@ bool CoupledDriver::is_converged()
-      CellHandle cell = elem_to_cell_.at(global_elem);
-      err_chk(heat.set_heat_source_at(local_elem, heat_source_.at(cell)),
-              "Error setting heat source for local element " +
-                std::to_string(local_elem));
+    if (relax) {
+      if (alpha_ == ROBBINS_MONRO) {
+        int n = i_picard_ + 1;
+        cell_heat_ = cell_heat_ / n + (1. - 1. / n) * cell_heat_prev_;
+      } else {
+        cell_heat_ = alpha_ * cell_heat_ + (1.0 - alpha_) * cell_heat_prev_;
+      }
+    }
+    for (gsl::index i = 0; i < cells_.size(); ++i) {
+      for (const auto& e : cell_to_elems_.at(cells_.at(i))) {
+        heat.set_heat_source_at(e, cell_heat_.at(i));
+      }
     }
   }
 }
 
 void CoupledDriver::update_temperature(bool relax)
 {
   comm_.message("Updating temperature");
-
   auto& neutronics = this->get_neutronics_driver();
   auto& heat = this->get_heat_driver();
 
-  // *************************************************************************
-  // Gather temperature on heat root and apply underrelaxation
-  // *************************************************************************
-
-  if (relax && comm_.rank == heat_root_) {
-    // Store previous temperature solution; a previous solution will always be present
-    // because a temperature IC is set and the neutronics solver runs first
-    std::copy(temperatures_.begin(), temperatures_.end(), temperatures_prev_.begin());
+  // Step 1: On each heat rank, assign the current iterate of local cell-avged T
+  // to the previous iterate
+  if (relax && heat.active()) {
+    std::copy(cell_temperatures_.begin(),
+              cell_temperatures_.end(),
+              cell_temperatures_prev_.begin());
   }
 
-  temperatures_ = heat.temperature();
-
-  if (relax && comm_.rank == heat_root_) {
-    if (alpha_T_ == ROBBINS_MONRO) {
-      int n = i_picard_ + 1;
-      temperatures_ = temperatures_ / n + (1. - 1. / n) * temperatures_prev_;
-    } else {
-      temperatures_ = alpha_T_ * temperatures_ + (1.0 - alpha_T_) * temperatures_prev_;
+  // Step 2: On each heat, compute cell-avged T
+  if (heat.active()) {
+    auto elem_temperatures = heat.temperature_local();
+    for (gsl::index i = 0; i < cells_.size(); ++i) {
+      double T_avg = 0.0;
+      double V_tot = 0.0;
+      for (const auto& e : cell_to_elems_.at(cells_.at(i))) {
+        double T = elem_temperatures.at(e);
+        double V = elem_volumes_.at(e);
+        T_avg += T * V;
+        V_tot += V;
+      }
+      T_avg /= V_tot;
+      Ensures(T_avg > 0.0);
+      cell_temperatures_.at(i) = T_avg;
+    }
+    // Apply relaxation to local cell-avged T
+    if (relax) {
+      if (alpha_T_ == ROBBINS_MONRO) {
+        int n = i_picard_ + 1;
+        cell_temperatures_ =
+          cell_temperatures_ / n + (1. - 1. / n) * cell_temperatures_prev_;
+      } else {
+        cell_temperatures_ =
+          alpha_T_ * cell_temperatures_ + (1.0 - alpha_T_) * cell_temperatures_prev_;
+      }
```

Comment:
To me, it's a little weird to be applying relaxation to the local cell averages, since those are only intermediate values that are not actually used to update temperatures. Is it not possible to do the relaxation after we already have the full cell average on the neutronics side?

Reply:
That makes a lot of sense and should be a simple fix.

Reply:
This actually touches a lot of code.  It gives us the opportunity to clean up some other stuff, and it's looks great so far.  But it would require a pretty extensive re-review.  Can I submit those updates in another PR?

Reply:
Yes, that's fine with me. Let's not mess up something we know is working :)